test/tls-provider.c

```diff
@@ -748,89 +1274,2261 @@ static const OSSL_DISPATCH xor_keymgmt_functions[] = {
+    xpk = xorx_key_to_pubkey(key, key_nid, str, strtype, k2d);
+
+    if (xpk != NULL)
+        ret = PEM_write_bio_X509_PUBKEY(out, xpk);
+    else
+        free_asn1_data(strtype, str);
+
+    /* Also frees |str| */
+    X509_PUBKEY_free(xpk);
+    return ret;
+}
+
+/* ---------------------------------------------------------------------- */
+
+static int prepare_xorx_params(const void *xorxkey, int nid, int save,
+                             void **pstr, int *pstrtype)
+{
+    ASN1_OBJECT *params = NULL;
+    XORKEY *k = (XORKEY*)xorxkey;
+
+    if (k->tls_name && OBJ_sn2nid(k->tls_name) != nid) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_INVALID_KEY);
+        return 0;
+    }
+
+    if (nid != NID_undef) {
+        params = OBJ_nid2obj(nid);
+        if (params == NULL)
+            return 0;
+    }
+    else {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_MISSING_OID);
+        return 0;
+    }
+
+
+    if (OBJ_length(params) == 0) {
+            /* unexpected error */
+            ERR_raise(ERR_LIB_USER, XORPROV_R_MISSING_OID);
+            ASN1_OBJECT_free(params);
+            return 0;
+    }
+    *pstr = params;
+    *pstrtype = V_ASN1_OBJECT;
+    return 1;
+}
+
+static int xorx_spki_pub_to_der(const void *vecxkey, unsigned char **pder)
+{
+    const XORKEY *xorxkey = vecxkey;
+    unsigned char *keyblob;
+    int retlen;
+
+    if (xorxkey == NULL) {
+        ERR_raise(ERR_LIB_USER, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    keyblob = OPENSSL_memdup(xorxkey->pubkey, retlen = XOR_KEY_SIZE);
+    if (keyblob == NULL) {
+        ERR_raise(ERR_LIB_USER, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    *pder = keyblob;
+    return retlen;
+}
+
+static int xorx_pki_priv_to_der(const void *vecxkey, unsigned char **pder)
+{
+    XORKEY *xorxkey = (XORKEY *)vecxkey;
+    unsigned char* buf = NULL;
+    ASN1_OCTET_STRING oct;
+    int keybloblen;
+
+    if (xorxkey == NULL) {
+        ERR_raise(ERR_LIB_USER, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    buf = OPENSSL_secure_malloc(XOR_KEY_SIZE);
+    memcpy(buf, xorxkey->privkey, XOR_KEY_SIZE);
+
+    oct.data = buf;
+    oct.length = XOR_KEY_SIZE;
+    oct.flags = 0;
+
+    keybloblen = i2d_ASN1_OCTET_STRING(&oct, pder);
+    if (keybloblen < 0) {
+        ERR_raise(ERR_LIB_USER, ERR_R_MALLOC_FAILURE);
+        keybloblen = 0;
+    }
+
+    OPENSSL_secure_clear_free(buf, XOR_KEY_SIZE);
+    return keybloblen;
+}
+
+# define xorx_epki_priv_to_der xorx_pki_priv_to_der
+
+/*
+ * XORX only has PKCS#8 / SubjectPublicKeyInfo
+ * representation, so we don't define xorx_type_specific_[priv,pub,params]_to_der.
+ */
+
+# define xorx_check_key_type     NULL
+
+# define xorhmacsig_evp_type       0
+# define xorhmacsig_input_type      XORSIGALG_NAME
+# define xorhmacsig_pem_type        XORSIGALG_NAME
+# define xorhmacsha2sig_evp_type       0
+# define xorhmacsha2sig_input_type      XORSIGALG_HASH_NAME
+# define xorhmacsha2sig_pem_type        XORSIGALG_HASH_NAME
```

Comment:
Can we line the right hand side of these macros up?

Reply:
done in https://github.com/openssl/openssl/pull/19312/commits/3cde11228ad9030f4d7efee78b60dd5bbea0ef0e